### PR TITLE
[13.0][MIG] product_gtin: Migration to 13.0

### PR DESCRIPTION
**Pull request by victoralmau**
Migration to 13.0

Please @joao-p-marques and @pedrobaeza can you review it?

@Tecnativa TT30724

### CODE REVIEW

File: product_gtin/models/abstract_barcodenumber.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# Copyright 2016 ACSONE SA/NV (<http://acsone.eu>)
+# Copyright 2021 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import barcodenumber
+from openerp import _, api, models
+from openerp.exceptions import ValidationError
+
+
+class AbstractBarcodenumber(models.AbstractModel):
+    _name = "abstract.barcodenumber"
```
Comment (pedrobaeza) [inline product_gtin/models/abstract_barcodenumber.py]: Use conventions:

```suggestion
    _name = "barcode.gtin_check.mixin"
```

File: product_gtin/models/abstract_barcodenumber.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# Copyright 2004-2011 Tiny SPRL (<http://tiny.be>)
+# Copyright 2010-2011 Camptocamp Austria (<http://www.camptocamp.at>)
+# Copyright 2016 ACSONE SA/NV (<http://acsone.eu>)
+# Copyright 2021 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import barcodenumber
+from openerp import _, api, models
```
Comment (pedrobaeza) [inline product_gtin/models/abstract_barcodenumber.py]: Change this to `odoo`

File: product_gtin/models/product_product.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+from openerp import api, models
+
+
+class ProductProduct(models.Model):
+    _name = "product.product"
+    _inherit = ["abstract.barcodenumber", "product.product"]
+    _barcodes = ["barcode"]
+
+    @api.constrains("barcode")
+    def _check_barecode(self):
```
Comment (pedrobaeza) [inline product_gtin/models/product_product.py]:
```suggestion
    def _check_barcode(self):
```

File: product_gtin/models/product_product.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+# Copyright 2016 ACSONE SA/NV (<http://acsone.eu>)
+# Copyright 2021 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openerp import api, models
+
+
+class ProductProduct(models.Model):
+    _name = "product.product"
+    _inherit = ["abstract.barcodenumber", "product.product"]
+    _barcodes = ["barcode"]
```
Comment (pedrobaeza) [inline product_gtin/models/product_product.py]: This is not serving for anything, but we can give it an utility:

- Move it to the mixin
- Put it a proper variable name:  `_barcode_fields = ["barcode"]` (or if it gets more complicated, do it a single field)
- Move the check to the mixin, and see if you can put on the `@api.constrains` the variable as arguments.
- On the inherited model, you only need to change the variable if not matching the field name in the model.

File: product_gtin/views/product_view.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+<openerp>
+    <data>
+        <record id="product_template_search_view" model="ir.ui.view">
+            <field name="model">product.template</field>
+            <field name="inherit_id" ref="product.product_template_search_view" />
+            <field name="arch" type="xml">
+                <field name="name" position="attributes">
+                    <attribute name="filter_domain">['|', '|',
+                        ('default_code','ilike',self),('name','ilike',self),
+                        ('barcode','ilike',self)]
```
Comment (pedrobaeza) [inline product_gtin/views/product_view.xml]: Isn't barcode already searchable in standard?

Comment (pedrobaeza) [reply product_gtin/views/product_view.xml]: Anyways, this is breaking any other inheritance.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: product_gtin/views/product_view.xml (hunk 5 above)
Comment (victoralmau) [reply product_gtin/views/product_view.xml]: It's not really necessary.

**Comment by victoralmau** [comment]
Thanks @pedrobaeza changes done.
Any idea to prevent incompatibility with other addons like `stock_barcodes_gs1`?. Maybe we need to add custom context `skip_barcode_gtin_check=True` in stock_barcodes_gs1 tests.

**Comment by pedrobaeza** [comment]
I think no error should be arisen if the length is different from the GTIN expected ones.

File: product_gtin/models/product_product.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2004-2011 Tiny SPRL (<http://tiny.be>)
+# Copyright 2010-2011 Camptocamp Austria (<http://www.camptocamp.at>)
+# Copyright 2016 ACSONE SA/NV (<http://acsone.eu>)
+# Copyright 2021 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openerp import models
```
Comment (joao-p-marques) [inline product_gtin/models/product_product.py]:
```suggestion
from odoo import models
```

File: product_gtin/tests/test_product_gtin_functions.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+class TestProductGtin(common.TransactionCase):
+    def setUp(self):
+        super().setUp()
+        self.barcode_mixin = self.env["barcode.gtin_check.mixin"]
+        self.product = self.env["product.product"].create({"name": "Test product"})
+
+    # The codes have been tested against
+    # http://www.hipaaspace.com/Medical_Data_Validation/Universal_Product_Code/
+    # UPC_Validation.aspx noqa
```
Comment (joao-p-marques) [inline product_gtin/tests/test_product_gtin_functions.py]: Why this `noqa` comment? Is it doing something important?
If so, could you explain so we can see if it worth fixing instead?

Also, the rest of the comments would probably fit better as a docstring of this method...

File: product_gtin/tests/test_product_gtin_functions.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        for code in VALID_UPC_CODES:
+            self.assertFalse(self.barcode_mixin._check_ean13(code))
+
+    def test_DICT_CHECK_CODE_TYPE(self):
+        """Check if the dict _DICT_CHECK_CODE_TYPE exists."""
+        self.assertTrue(self.barcode_mixin._DICT_CHECK_CODE_TYPE)
+
+    def test_DICT_CHECK_FUNCTIONS(self):
+        """Check if the dict _DICT_CHECK_FUNCTIONS exists."""
+        self.assertTrue(self.barcode_mixin._DICT_CHECK_FUNCTIONS)
```
Comment (joao-p-marques) [inline product_gtin/tests/test_product_gtin_functions.py]: You could probably concentrate some of these tests in a singe one, without the need for a function name so specific...
```suggestion
    def test_dicts_check_exist(self):
        """Check if the dicts _DICT_CHECK_CODE_TYPE and _DICT_CHECK_FUNCTIONS exist."""
        self.assertTrue(self.barcode_mixin._DICT_CHECK_CODE_TYPE)
        self.assertTrue(self.barcode_mixin._DICT_CHECK_FUNCTIONS)
```

The same applies to some of the other functions that just have 1 or 2 lines, if you think it is appropriate.

**Review by joao-p-marques — CHANGES_REQUESTED** [review]
Code review. Looks good, apart from this small comments :+1:

File: product_gtin/tests/test_product_gtin_functions.py (hunk 7 above)
Comment (victoralmau) [reply product_gtin/tests/test_product_gtin_functions.py]: It's not really necessary, i have removed it now.

File: product_gtin/tests/test_product_gtin_functions.py (hunk 8 above)
Comment (victoralmau) [reply product_gtin/tests/test_product_gtin_functions.py]: Thanks, i have improved some functions to simplify validations.

**Review by joao-p-marques — APPROVED** [review]
Code and functional review :+1:

**Review by CarlosRoca13 — APPROVED** [review]

**Review by pedrobaeza — APPROVED** [review]

**Comment by victoralmau** [comment]
It seems that the bot has locked, can you launch it again?

**Pull request merged**